src/CloudNative.CloudEvents/CloudEventsSpecVersion.cs

```diff
@@ -1,15 +1,146 @@
-﻿// Copyright (c) Cloud Native Foundation. 
+﻿// Copyright 2021 Cloud Native Foundation. 
 // Licensed under the Apache 2.0 license.
 // See LICENSE file in the project root for full license information.
 
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Mime;
+using static CloudNative.CloudEvents.CloudEventAttribute;
+
 namespace CloudNative.CloudEvents
 {
-    public enum CloudEventsSpecVersion
+    // TODO: Rename to CloudEventSpecVersion? Using the plural form feels a little odd.
+
+    /// <summary>
+    /// Represents a version of the CloudEvents specification, including
+    /// the context attribute values known to that version.
+    /// </summary>
+    public sealed class CloudEventsSpecVersion
     {
-        V0_1,
-        V0_2,
-        V0_3,
-        V1_0,
-        Default = V1_0
+        /// <summary>
+        /// The attribute used to indicate the version of the CloudEvents specification being used.
+        /// </summary>
+        public static CloudEventAttribute SpecVersionAttribute { get; } = CreateRequired("specversion", CloudEventAttributeType.String, NonEmptyString);
+
+        // Populated by the constructor.
+        private static readonly List<CloudEventsSpecVersion> allVersions = new List<CloudEventsSpecVersion>();
+
+        public static CloudEventsSpecVersion Default => V1_0;
+
+        public static CloudEventsSpecVersion V1_0 { get; } = new CloudEventsSpecVersion(
+            "1.0",
+            CreateRequired("id", CloudEventAttributeType.String, NonEmptyString),
```

Comment:
nit: maybe calculate the validator based on the attribute type within `CreateRequired`.

Reply:
No, the type validates the content separately. This is additional semantic validation. So empty strings are generally fine as context attributes, but they're *not* fine for the id attribute.

Reply:
but doesn't calling CreateRequired already specify that the attribute can't be empty?

Reply:
Oh, I see what you mean. I don't think there's anything technically stopping a required attribute from being empty - it's just that all the current required attributes also have a non-empty requirement :)

Given how few callers there will be to CreateRequired (it's internal) I think I'd prefer to be explicit here.